### DISCUSSION TITLE
how to Interface TDEP code

### DISCUSSION THREAD

**Discussion by Liangzhang545** [phoebe-team/phoebe#183] [Q&A] [answered]
Dear Jenny
I am Liang Zhang from China. There are two questions that confuse me.
1. Electric transport, can you replace the second-order force constant of QE with the 2rd force constant of TDEP to study the effect of phonon reforming on electric transport?
2. Thermal transport, can you directly read the 2rd, 3rd and 4th-order force constants of TDEP to study the effect of phonon reforming on thermal transport?
TDEP code website: https://ollehellman.github.io/page/workflows/minimal_example_5.html
I sincerely look forward to your reply
yours
Liang Zhang

**Comment by jcoulter12** [comment]
Hi Liang Zhang, 

Thanks for your interest in Phoebe!

As for your questions: 

> 1) Electric transport, can you replace the second-order force constant of QE with the 2rd force constant of TDEP to study the effect of phonon reforming on electric transport?

If you have force constants in the .fc format of QE, you could supply that to Phoebe and the calculation should run. However, would you also need some correction to the electron-phonon matrix elements? This is likely a physics question you need to consider a bit further. 

> 2) Thermal transport, can you directly read the 2rd, 3rd and 4th-order force constants of TDEP to study the effect of phonon reforming on thermal transport?

Currently, Phoebe can read 3rd-order force constants from phono3py or ShengBTE (when ShengBTE is used with QE). 
We're open to the possibility of interfacing with other codes, however, and it would be interesting to interface with TDEP. 
I'm not very familiar with their file formats -- it looks like they use their own filetype? 

If that's the case, I'd have to write a new parser for the format. It might take me a little while to get to adding this, but if you can send me a documentation page describing their file format, I can let you know if it's possible for me to do.

We'd love to someday also add the 4th order force constants, but that feature is a bit more of an endeavor to add, so unfortunately I don't have a timescale as to when in the future we might offer this. 

Let me know if you have any more questions, 
Jenny

**Comment by Liangzhang545** [comment]
有兴趣的朋友可以去试下!每天赚几十元,可以~当额外收入~*详情请登陆招聘网站http://hjmqtv.cn/

**Reply by jcoulter12** [reply]
I'm not sure if you meant to post a reply here. :)

**Comment by Liangzhang545** [comment]
Dear Sir
I am very grateful for your reply. I am not very familiar with this TDEP software either. According to its official website (https://ollehellman.github.io/page/index.html), this software can output second-order, third-order, and fourth-order force constants. Based on the original package and these force constants, TDEP can generate thermal conductivity information. ShengBTE can also access these files to generate thermal conductivity information. Therefore, I hope that phoebe software can also access these input files like ShengBTE to generate thermal conductivity information.&nbsp;I am very looking forward to the implementation of these new features. Look forward to your reply.

Yours&nbsp;
Liang Zhang



建联
***@***.***







                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/9 20:36

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Author"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
Hi Liang Zhang,
 
Thanks for your interest in Phoebe!
 
As for your questions:
  
Electric transport, can you replace the second-order force constant of QE with the 2rd force constant of TDEP to study the effect of phonon reforming on electric transport?
  
If you have force constants in the .fc format of QE, you could supply that to Phoebe and the calculation should run. However, would you also need some correction to the electron-phonon matrix elements? This is likely a physics question you need to consider a bit further.
  
Thermal transport, can you directly read the 2rd, 3rd and 4th-order force constants of TDEP to study the effect of phonon reforming on thermal transport?
  
Currently, Phoebe can read 3rd-order force constants from phono3py or ShengBTE (when ShengBTE is used with QE).
 We're open to the possibility of interfacing with other codes, however, and it would be interesting to interface with TDEP.
 I'm not very familiar with their file formats -- it looks like they use their own filetype?
 
If that's the case, I'd have to write a new parser for the format. It might take me a little while to get to adding this, but if you can send me a documentation page describing their file format, I can let you know if it's possible for me to do.
 
We'd love to someday also add the 4th order force constants, but that feature is a bit more of an endeavor to add, so unfortunately I don't have a timescale as to when in the future we might offer this.
 
Let me know if you have any more questions,
 Jenny

**Comment by Liangzhang545** [comment]
Dear Jenny,For electric transport and thermal transport, the essential calculation is phonon part, but this process is usually calculated at 0K and does not consider the temperature effect. I think that the temperature-induced anharmonic phonon renormalization (APRN) is important, so I want to study the effect of APRN on the electric transport and thermal transport.
At now, there is no code, which could study the electric transport under APRN. In order to address this problem, the 2nd- order IFC of QE is need to be replaced by the 2nd- order IFC of TDEP. If the phoebe implements this function, there will be more people to use phoebe, because APRN is a frontier research.
For thermal transport, although the TDEP could produce the 2rd, 3rd and 4th-order IFCs, the thermal conductivity does not include the contribution of four-phonon scattering by TDEP code. As far as I know, there is no software that can simultaneously consider 2rd, 3rd and 4th-order IFCs under APRN to calculate the lattice thermal conductivity.
In addition, for 2D material, Does the phoebe program consider the thickness and cut-off phonon frequency of two-dimensional materials? Because the ZA branch of 2D material is difficult to deal with and leads to the divergence of mobility.
The attachment is the force constant of TDEP, including 2rd, 3rd and 4th-order IFCs.
Thank you very much.




建联
***@***.***







                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/9 20:36

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Author"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
Hi Liang Zhang,
 
Thanks for your interest in Phoebe!
 
As for your questions:
  
Electric transport, can you replace the second-order force constant of QE with the 2rd force constant of TDEP to study the effect of phonon reforming on electric transport?
  
If you have force constants in the .fc format of QE, you could supply that to Phoebe and the calculation should run. However, would you also need some correction to the electron-phonon matrix elements? This is likely a physics question you need to consider a bit further.
  
Thermal transport, can you directly read the 2rd, 3rd and 4th-order force constants of TDEP to study the effect of phonon reforming on thermal transport?
  
Currently, Phoebe can read 3rd-order force constants from phono3py or ShengBTE (when ShengBTE is used with QE).
 We're open to the possibility of interfacing with other codes, however, and it would be interesting to interface with TDEP.
 I'm not very familiar with their file formats -- it looks like they use their own filetype?
 
If that's the case, I'd have to write a new parser for the format. It might take me a little while to get to adding this, but if you can send me a documentation page describing their file format, I can let you know if it's possible for me to do.
 
We'd love to someday also add the 4th order force constants, but that feature is a bit more of an endeavor to add, so unfortunately I don't have a timescale as to when in the future we might offer this.
 
Let me know if you have any more questions,
 Jenny

**Reply by jcoulter12** [reply]
Hi @Liangzhang545, 

I have a few parts to this reply: 
>At now, there is no code, which could study the electric transport under APRN.

I have a feeling this is because, to my knowledge, there's no way to temperature correct the electron-phonon matrix elements calculated using DFT. While you could correct the phonon frequencies, and maybe this would approximate some of the effects of temperature, how would one consider the change of elph interactions? Have you considered if this calculation could every actually produce a physically meaningful result? 

>For thermal transport, although the TDEP could produce the 2rd, 3rd and 4th-order IFCs, the thermal conductivity does not include the contribution of four-phonon scattering by TDEP code.

I'm investingating if there's a way I can reformat the output of TDEP into either the format of ShengBTE or Phono3py. If this were the case, I could add the capability to use the 2nd and 3rd order force constants hopefully without too much of a wait time. We would like to add 4th order phonon scattering as well, however, we currently have other developments that are higher priority for us. 

>In addition, for 2D material, Does the phoebe program consider the thickness and cut-off phonon frequency of two-dimensional materials? Because the ZA branch of 2D material is difficult to deal with and leads to the divergence of mobility.

Can you give me an example of what you mean by cut off phonon frequency? Currently, if you run a calculation with an unstable phonon mode, we throw a warning and discard the states with negative frequencies to prevent the computation from outputting garbage. Is this the kind of thing you are referring to? As for thickness, yes, when you set dimensionality=2, we consider the thickness of the material. 

For now, if you can send me your 2nd and 3rd order force constants, that is somewhat helpful, though I may need other information about the calculation. You can send them to my email address at jcoulter@g.harvard.edu, or share a link to a google drive folder. If there's a future time when we are also considering the 4th order force constants, I'll let you know. 

Best, 
Jenny

**Comment by Liangzhang545** [comment]
primitivecell file.




建联
***@***.***







                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/10 16:50

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Author"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
I'm not sure if you meant to post a reply here. :)

**Comment by Liangzhang545** [comment]
infile.forceconstant_thirdorder




建联
***@***.***







                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/10 16:50

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Author"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
I'm not sure if you meant to post a reply here. :)

**Comment by Liangzhang545** [comment]
infile.forceconstant_fourthorder


                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/10 16:50

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Author"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
I'm not sure if you meant to post a reply here. :)

**Comment by Liangzhang545** [comment]
My email size exceeds the limit of the recipient's service provider and cannot be delivered.&nbsp;Can you provide another email address for me to transfer the remaining files to you ？




建联
***@***.***







                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/10 16:50

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Author"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
I'm not sure if you meant to post a reply here. :)

**Comment by jcoulter12** [comment]
@Liangzhang545, something else I would likely need is access to TDEP, which seems to be a private repository? 
It's much less favorable for us to interface with a code that isn't even available to the public.

**Comment by Liangzhang545** [comment]
Dear Jenny,&nbsp; &nbsp; &nbsp;For electric transport, I want to consider the temperature effect by correcting the phonon frequency, and the 2nd- order IFC of QE is need to be replaced by the 2nd- order IFC of TDEP.
&nbsp; &nbsp; &nbsp;For thermal transport, this is the link of TDEP, https://github.com/Npikeulg/TDEP_copy, could you open this website. Combining the output of TDEP and ph-el scattering, I think this is perfect.
&nbsp; &nbsp; &nbsp;For 2D, about phfreq_cutoff (meV), it means that phonons with energy is smaller than phfreq_cutoff will be excluded.
&nbsp; &nbsp; &nbsp;I have send the 2nd and 3rd order force constants to your email address at ***@***.***




建联
***@***.***







                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/15 17:59

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Mention"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
@Liangzhang545, something else I would likely need is access to TDEP, which seems to be a private repository?
 It's much less favorable for us to interface with a code that isn't even available to the public.

**Comment by Liangzhang545** [comment]
Dear Sir
I met &nbsp;multiple q-points report errors when&nbsp;calculating thermal conductivity. It is OK when q point is 1*1*1. It return error when&nbsp;q point is 2*2*1 or even densier q point mesh. The attach files are input files





建联
***@***.***







                       
Original Email
                       
                     

Sender:"Jenny Coulter"< ***@***.*** &gt;;

Sent Time:2023/5/15 17:59

To:"mir-group/phoebe"< ***@***.*** &gt;;

Cc recipient:"Liangzhang545"< ***@***.*** &gt;;"Mention"< ***@***.*** &gt;;

Subject:Re: [mir-group/phoebe] how to Interface TDEP code (Discussion #183)




 
@Liangzhang545, something else I would likely need is access to TDEP, which seems to be a private repository?
 It's much less favorable for us to interface with a code that isn't even available to the public.

**Comment by jcoulter12** [comment]
Hi @Liangzhang545, 

I don't see any attached input files, unfortunately. 
I would recommend using the normal web interface to github to upload the files, I get the idea that you're using some client to send these messages because of all the extra characters it's adding, I would suspect this is part of the issue with your file uploads. 

Thanks, 
Jenny

**Comment by Liangzhang545** [comment]
有兴趣的朋友可以去试下!每天赚几十元,可以~当额外收入~*详情请登陆招聘网站http://hjmqtv.cn/

**Reply by jcoulter12** [reply]
Again, I don't think you meant to send this.

**Comment by jcoulter12** [comment, marked as answer]
@Liangzhang545 -- 

I just thought again of this discussion because I realized temperature-dependent phonon force constants can be obtained using the SSCHA method, which can write them to the QE .fc format. 

If you wanted to use temperature dependent phonons in a calculation, Phoebe natively takes .fc as a file format. 

Just wanted to post this here as well in case it's relevant for someone else later. 

Thanks, 
Jenny

**Comment by Liangzhang545** [comment]
有兴趣的朋友可以去试下!每天赚几十元,可以~当额外收入~*详情请登陆招聘网站http://hjmqtv.cn/
